### DISCUSSION TITLE
Update Existing SubstructLibrary

### DISCUSSION THREAD

**Discussion by sudoPete** [rdkit/rdkit#5019] [Q&A] [answered]
I am trying to update an existing substructure library, however, it's giving me an attribute error when I try to call AddSmiles. Is it possible to update a rdSubstructLibrary after it has been initialized? example of code that errors below:

```python
from rdkit import Chem
from rdkit.Chem import rdSubstructLibrary

mols = rdSubstructLibrary.CachedTrustedSmilesMolHolder()
fps = rdSubstructLibrary.PatternHolder()
smiles = ["C", "CC", "CCCC"]
for smile in smiles:
    fp = Chem.PatternFingerprint(Chem.MolFromSmiles(smile))
    mols.AddSmiles(smile)
    fps.AddFingerprint(fp)
library = rdSubstructLibrary.SubstructLibrary(mols, fps)
new_smiles = "CCCCC"
library.AddSmiles(new_smiles)
fp = Chem.PatternFingerprint(Chem.MolFromSmiles(new_smiles))
library.addPatterns(fp)
```

**Comment by bp-kelley** [comment]
The Library API doesn't know what holder you chose to use so you'll need to
use the AddMol function:

library.AddMol(Chem.MolFromSmiles(new_smiles))

This will also properly add the fingerprints as well if you are using them.

The use case for AddSmiles and AddFingerprints is if you already have the
data cached, it is much, much faster to load.

Cheers,
 Brian

**Reply by sudoPete** [reply]
Thanks for the reply!

In the real code base smiles, mol and fingerprint are already cached, regardless, I still need to use add mol function after initialization?

**Comment by bp-kelley** [comment, marked as answer]
In that case, just add the data to the original molecule and pattern holders

```
mols.AddSmiles(smile)
fps.AddFingerprint(fp)
```

These are also available from the library itself:

```
>>> library.GetMolHolder()
<rdkit.Chem.rdSubstructLibrary.CachedTrustedSmilesMolHolder object at 0x102f289a0>
>>> library.GetFpHolder()
<rdkit.Chem.rdSubstructLibrary.PatternHolder object at 0x102b1da90>
```

Cheers,
  Brian

**Comment by bp-kelley** [comment]
But then the book-keeping is entirely up to you :)
